kaldb/src/main/java/com/slack/kaldb/preprocessor/PreprocessorRateLimiter.java

```diff
@@ -102,6 +105,86 @@ protected static RateLimiter smoothBurstyRateLimiter(
     }
   }
 
+  static class RateLimiterPerDatasetHolder {
+    public DatasetMetadata datasetMetadata;
+    public RateLimiter rateLimiter;
+
+    public RateLimiterPerDatasetHolder(DatasetMetadata datasetMetadata, RateLimiter rateLimiter) {
+      this.datasetMetadata = datasetMetadata;
+      this.rateLimiter = rateLimiter;
+    }
+  }
+
+  public Predicate<String, Trace.Span> createRateLimiter(
+      List<DatasetMetadata> datasetMetadataList) {
+
+    List<DatasetMetadata> throughputSortedDatasets =
+        datasetMetadataList
+            .stream()
+            .sorted(Comparator.comparingLong(DatasetMetadata::getThroughputBytes))
+            .collect(Collectors.toList());
+
+    List<RateLimiterPerDatasetHolder> rateLimiterPerDatasetList = new ArrayList<>();
+    for (DatasetMetadata datasetMetadata : throughputSortedDatasets) {
+      double permitsPerSecond = (double) datasetMetadata.getThroughputBytes() / preprocessorCount;
+      LOG.info(
+          "Rate limiter initialized for {} at {} bytes per second (target throughput {} / processorCount {})",
+          datasetMetadata.getName(),
+          permitsPerSecond,
+          datasetMetadata.getThroughputBytes(),
+          preprocessorCount);
+      RateLimiter rateLimiter =
+          smoothBurstyRateLimiter(permitsPerSecond, maxBurstSeconds, initializeWarm);
+      rateLimiterPerDatasetList.add(new RateLimiterPerDatasetHolder(datasetMetadata, rateLimiter));
+    }
+
+    return (key, value) -> {
+      if (value == null) {
+        LOG.warn("Message was dropped, was null span");
+        return false;
+      }
+
+      String serviceName = PreprocessorValueMapper.getDatasetName(value);
+      int bytes = value.getSerializedSize();
+      if (serviceName == null || serviceName.isEmpty()) {
+        // service name wasn't provided
+        LOG.debug("Message was dropped due to missing service name - '{}'", value);
+        // todo - we may consider adding a logging BurstFilter so that a bad actor cannot
+        //  inadvertently swamp the system if we want to increase this logging level
+        //  https://logging.apache.org/log4j/2.x/manual/filters.html#BurstFilter
+        meterRegistry
+            .counter(MESSAGES_DROPPED, getMeterTags("", MessageDropReason.MISSING_SERVICE_NAME))
+            .increment();
+        meterRegistry
+            .counter(BYTES_DROPPED, getMeterTags("", MessageDropReason.MISSING_SERVICE_NAME))
+            .increment(bytes);
+        return false;
+      }
+
+      for (RateLimiterPerDatasetHolder rateLimiterPerDataset : rateLimiterPerDatasetList) {
```

Comment:
Why are we iterating through this tuple? Can't we just iterate through the list of datasets to find the matching one, and then get the matching rate limiter from a map, keyed against dataset? It seems like the tuple shouldn't be necessary?